src/client-agent/start_agent.c

```diff
@@ -166,134 +169,193 @@ void start_agent(int is_startup)
-                            sleep(agt->enrollment_cfg->delay_after_enrollment);
-                            // Successfull enroll, read keys
-                            OS_UpdateKeys(&keys);
-                        }
-                    }
-                    if (!connect_server(agt->rip_id)) {
-                        continue;
-                    }
+            /* Send message again (after three attempts) */
+            if (attempts >= 3 || recv_b == OS_SOCKTERR) {
+                if (attempts == 3 && agt->enrollment_cfg && agt->enrollment_cfg->enabled) { // Only one enrollment attemp
+                    try_enroll_to_server(agt->server[agt->rip_id].rip);
+                }
+                if (connect_server(agt->rip_id)) {
                     // if enroll is successfull reconnect and re-send message
                     send_msg(msg, -1);
                     // After sending message wait before response
-                    if (agt->server[agt->rip_id].protocol == IPPROTO_UDP) {
-                        sleep(attempts);
-                    }
+                    sleep(attempts);
                 }
-
-                continue;
-            }
-
-            /* Id of zero -- only one key allowed */
-            if (ReadSecMSG(&keys, buffer, cleartext, 0, recv_b - 1, &msg_length, agt->server[agt->rip_id].rip, &tmp_msg) != KS_VALID) {
-                mwarn(MSG_ERROR, agt->server[agt->rip_id].rip);
-                continue;
             }
+        } else {
+            return recv_b;
+        }   
+    }
+    return 0;
+}
 
-            /* Check for commands */
-            if (IsValidHeader(tmp_msg)) {
-                /* If it is an ack reply */
-                if (strcmp(tmp_msg, HC_ACK) == 0) {
-                    available_server = time(0);
-
-                    minfo(AG_CONNECTED, agt->server[agt->rip_id].rip,
-                            agt->server[agt->rip_id].port, agt->server[agt->rip_id].protocol == IPPROTO_UDP ? "udp" : "tcp");
-
-                    if (is_startup) {
-                        /* Send log message about start up */
-                        snprintf(msg, OS_MAXSTR, OS_AG_STARTED,
-                                 keys.keyentries[0]->name,
-                                 keys.keyentries[0]->ip->ip);
-                        snprintf(fmsg, OS_MAXSTR, "%c:%s:%s", LOCALFILE_MQ,
-                                 "ossec", msg);
-                        send_msg(fmsg, -1);
+/**
+ * Holds the message reception logic for tcp
+ * @param msg message to be sent
+ * @param buffer pointer to buffer where the information will be stored
+ * @param max_length size of buffer
+ * @return 1 on success
+ *         0 when all retries failed
+ * */
+static ssize_t receive_message_tcp(const char *msg, char *buffer, unsigned int max_lenght) {
+    ssize_t recv_b = 0;
+    int attempts = 0;
+    bool enrollment_attemp = false;
+    
+    while ((attempts < 5) && (recv_b <= 0)) {
```

Comment:
Previously, we wait an incremental time between retries (`sleep(attempts);`), is now unnecesary?

Reply:
This seems necessary for the UDP case because it is not blocking but TCP socket blocks until a configurable timeout is reached. We can discuss with @vikman90  whetever we should or not increment this timeout in case of failures